django/db/backends/base/schema.py

```diff
@@ -767,19 +721,72 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
         # Drop the default if we need to
         # (Django usually does not use in-database defaults)
         if needs_database_default:
-            sql = self.sql_alter_column % {
+            self.execute(*self._alter_column_default_sql(model, old_field, new_field, True))
+        # Reset connection if required
+        if self.connection.features.connection_persists_old_columns:
+            self.connection.close()
+
+    def _alter_column_null_sql(self, model, old_field, new_field):
+        """
+        Hook to specialize column null alteration for different backends.
+
+        Return a sql fragment (sql, params).
+        """
+        new_db_params = new_field.db_parameters(connection=self.connection)
+        new_type = new_db_params['type']
+
+        if (self.connection.features.interprets_empty_strings_as_nulls and
+                new_field.get_internal_type() in ("CharField", "TextField")):
+            # The field is nullable in the database anyway, leave it alone
+            return
+        else:
+            sql = self.sql_alter_column_null if new_field.null else self.sql_alter_column_not_null
+            return (
+                sql % {
+                    "column": self.quote_name(new_field.column),
+                    "type": new_type,
+                },
+                [],
+            )
+
+    def _alter_column_default_sql(self, model, old_field, new_field, drop=False):
```

Comment:
I am currently not really happy with the fact that this method uses `self.sql_alter_column` (ie returns the full statement) in case of dropping the default and a partial one otherwise, I think I am going to get rid of `sql_alter_column` here and do it at the calling site instead.